x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/rule_types/create_security_rule_type_wrapper.ts

```diff
@@ -252,6 +283,10 @@ export const createSecurityRuleTypeWrapper: CreateSecurityRuleTypeWrapper =
             }
           }
 
+          // Make a copy of `inputIndex` or else the APM agent reports it as [Circular] for most rule types because it's the same object
```

Comment:
I won't disclose how many times I've read this sentence, but: I still don't quite understand what's going on here. It looks like it's caused by [this line](https://github.com/elastic/kibana/blob/main/x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/rule_types/create_security_rule_type_wrapper.ts#L238), and the agent doesn't like two keys referencing the same object?

Or: is there a bug in the agent, where it's e.g. trying to make its own copy of the context and gets confused on this?

Reply:
> the agent doesn't like two keys referencing the same object?

This, the agent keeps track of objects that have already been set in the context in an attempt to avoid un-serializable objects. Since the `inputIndex` also lives in the rule params if the rule is not using data views, the same object could be inserted twice unless we make a copy here because we include both the full rule params and the resolved `inputIndex`.

We could theoretically include the explicit input index here conditionally, but I think it's more useful to have the context be as consistent as possible for various rule configurations and always have the input index defined here.

Reply:
Makes sense.